src/contract.rs

```diff
@@ -914,13 +999,120 @@ fn query_validators_metrics_by_timestamp(
     Ok(res)
 }
 
+
+// off chain code
+
+//todo: test
+fn remove_off_chain_metrics_for_timestamp(deps: DepsMut, info: MessageInfo, timestamp: u64)
+                                          -> Result<Response, ContractError> {
+
+    if !sender_is_manager(&deps, &info) {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let mut off_chain_timestamps = OFF_CHAIN_TIMESTAMPS.load(deps.storage)?;
+
+    let existing_timestamps_length = off_chain_timestamps.timestamps.len();
+
+    let new_timestamps = off_chain_timestamps.timestamps.into_iter()
+        .filter(|state_timestamp| { state_timestamp.ne(&timestamp)})
+        .collect::<Vec<u64>>();
+
+    let mut timestamp_existed = false;
+
+    if new_timestamps.len() < existing_timestamps_length {
+        off_chain_timestamps.timestamps = new_timestamps;
+        OFF_CHAIN_TIMESTAMPS.save(deps.storage, &off_chain_timestamps)?;
+        OFF_CHAIN_TIMESTAMP_DETAILS.remove(deps.storage, U64Key::from(timestamp));
+        timestamp_existed = true;
+    }
+
+    Ok(Response::new()
+        .add_attribute("method", "remove_off_chain_metrics_for_timestamp")
+        .add_attribute("message",
+                       if timestamp_existed {"timestamp successfully removed"}
+                       else {"timestamp didn't exist"}))
+}
+
+fn add_off_chain_validator_metrics(deps: DepsMut,
+                                   info: MessageInfo,
+                                   timestamp: u64,
+                                   metrics_to_be_added: Vec<OffChainValidatorMetrics>)
+    -> Result<Response, ContractError> {
+
+    if !sender_is_manager(&deps, &info) {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    for validator_metric in metrics_to_be_added {
+        if off_chain_metrics_exists(&deps, timestamp, &validator_metric.opr_address) {
+            return Err(ContractError::OffChainMetricsAlreadyRecorded)
+        }
+        OFF_CHAIN_STATE_FOR_VALIDATOR.save(
+            deps.storage,
+            (U64Key::from(timestamp), validator_metric.opr_address.clone()),
```

Comment:
This should be (U64Key::from(timestamp), validator_index).

Get the validator index from the other mapping.